src/hotspot/share/cds/aotReferenceObjSupport.cpp

```diff
@@ -0,0 +1,241 @@
+//     is eligible.
+//
+// [2] A reference that REQUIRE specials clean up (i.e., Reference::queue != _null_queue.resolve())
+//     is eligible ONLY if its referent is not null.
+//
+// As of this version, the only oops in group [2] that can be found by AOTArtifactFinder are
+// the keys used by ReferencedKeyMap in the implementation of MethodType::internTable.
+// stabilize_cached_reference_objects() ensures that all keys found by AOTArtifactFinder are eligible.
+//
+// The purpose of the error check in check_if_ref_obj() is to guard against changes in the JDK core
+// libs that might introduce new types of oops in group [2] into the AOT cache.
+//
+// Reasons for the eligibility restrictions
+// ========================================
+//
+// Reference handling is complex. In this version, we implement only enough functionality to support
+// the use of Weak/Soft references used by java.lang.invoke.
+//
+// We intend to evolve the implementation in the future by
+// -- implementing more assemblySetup() operations for other use cases, and/or
+// -- relaxing the eligibility restrictions.
+//
+//
+// null referents for group [1]
+// ============================
+//
+// Any cached reference R1 of group [1] is allowed to have a null referent.
+// This can happen in the following situations:
+//    (a) R1.clear() was called by Java code during the assembly phase.
+//    (b) The referent has been collected, and R1 is in the "pending" state.
+// In case (b), the "next" and "discovered" fields of the cached copy of R1 will
+// be set to null. During the production run:
+//    - It would appear to the Java program as if immediately during VM start-up, the referent
+//      was collected and ReferenceThread completed processing of R1.
+//    - It would appear to the GC as if immediately during VM start-up, the Java program called
+//      R1.clear().
+
+#if INCLUDE_CDS_JAVA_HEAP
+
+
+class KeepAliveObjectsTable : public ResourceHashtable<oop, bool,
+    36137, // prime number
+    AnyObj::C_HEAP,
+    mtClassShared,
+    HeapShared::oop_hash> {};
+
+static KeepAliveObjectsTable* _keep_alive_objs_table;
+static OopHandle _keep_alive_objs_array;
+static OopHandle _null_queue;
+
+bool AOTReferenceObjSupport::is_enabled() {
+  // For simplicity, AOTReferenceObjSupport is enabled only when dumping method handles.
+  // Otherwise we won't see Reference objects in the AOT cache. Let's be conservative now.
+  return CDSConfig::is_dumping_method_handles();
+}
+
+void AOTReferenceObjSupport::initialize(TRAPS) {
+  if (!AOTReferenceObjSupport::is_enabled()) {
+    return;
+  }
+
+  TempNewSymbol class_name = SymbolTable::new_symbol("java/lang/ref/ReferenceQueue");
+  Klass* k = SystemDictionary::resolve_or_fail(class_name, true, CHECK);
+  InstanceKlass* ik = InstanceKlass::cast(k);
+  ik->initialize(CHECK);
+
+  TempNewSymbol field_name = SymbolTable::new_symbol("N""ULL");
```

Comment:
Why "N""ULL", not "NULL"?

Reply:
That's because a stand-lone word "NULL" will cause hotspot/jtreg/sources/TestNoNULL.java to fail. I have renamed the internal field to NULL_QUEUE to avoid the failure, per recommendation of @kimbarrett

Reply:
Thank you for the explanation!